File: packages/core/src/checksum.js
// FNV-1a 32-bit over raw Int32 state.
//
// The hash walks the state buffer in index order and feeds each Int32 as four
// explicit little-endian bytes. It never views the underlying ArrayBuffer as
// bytes, so a big-endian host produces the same digest as a little-endian one,
// and the digest is stable as long as the field layout is stable.

const FNV_OFFSET = 0x811c9dc5;
const FNV_PRIME = 0x01000193;

/** FNV-1a over an array-like of Int32 values. Returns an unsigned 32-bit int. */
export function fnv1aInts(ints, seed = FNV_OFFSET) {
  let h = seed >>> 0;
  for (let i = 0; i < ints.length; i++) {
    const v = ints[i] | 0;
    h = (Math.imul(h ^ (v & 0xff), FNV_PRIME) >>> 0);
    h = (Math.imul(h ^ ((v >>> 8) & 0xff), FNV_PRIME) >>> 0);
    h = (Math.imul(h ^ ((v >>> 16) & 0xff), FNV_PRIME) >>> 0);
    h = (Math.imul(h ^ ((v >>> 24) & 0xff), FNV_PRIME) >>> 0);
  }
  return h >>> 0;
}

/** FNV-1a over the UTF-16 code units of a string, byte-wise little-endian. */
export function fnv1aString(str, seed = FNV_OFFSET) {
  let h = seed >>> 0;
  for (let i = 0; i < str.length; i++) {
    const c = str.charCodeAt(i);
    h = (Math.imul(h ^ (c & 0xff), FNV_PRIME) >>> 0);
    h = (Math.imul(h ^ ((c >>> 8) & 0xff), FNV_PRIME) >>> 0);
  }
  return h >>> 0;
}

/** Unsigned 32-bit int -> lowercase 8-digit hex. */
export function toHex32(h) {
  return (h >>> 0).toString(16).padStart(8, '0');
}

/** Convenience: hex digest of an Int32 array-like. */
export function checksumInts(ints) {
  return toHex32(fnv1aInts(ints));
}

File: python/haxball3d_sim/checksum.py
"""FNV-1a 32-bit, mirroring packages/core/src/checksum.js.

The JS hash walks an Int32 array and feeds each value as four explicit
little-endian bytes; it never views the underlying ArrayBuffer, so the digest
is host-endianness independent and this port only has to reproduce the byte
order, not the memory layout.

``Math.imul(h, PRIME) >>> 0`` is a 32-bit multiply that keeps the low 32 bits,
so the Python form is ``(h * PRIME) & 0xFFFFFFFF``. The XOR in JS runs through
ToInt32 first, but XOR and multiplication agree modulo 2**32 whichever sign
convention is used, so the unsigned form here is exact.
"""

from __future__ import annotations

import numpy as np

FNV_OFFSET = 0x811C9DC5
FNV_PRIME = 0x01000193
_MASK = 0xFFFFFFFF


def fnv1a_ints(ints, seed: int = FNV_OFFSET) -> int:
    """FNV-1a over a sequence of Int32 values. Returns an unsigned 32-bit int."""
    h = seed & _MASK
    for value in ints:
        v = int(value) & _MASK
        h = ((h ^ (v & 0xFF)) * FNV_PRIME) & _MASK
        h = ((h ^ ((v >> 8) & 0xFF)) * FNV_PRIME) & _MASK
        h = ((h ^ ((v >> 16) & 0xFF)) * FNV_PRIME) & _MASK
        h = ((h ^ ((v >> 24) & 0xFF)) * FNV_PRIME) & _MASK
    return h


def fnv1a_bytes(data: bytes, seed: int = FNV_OFFSET) -> int:
    h = seed & _MASK
    for byte in data:
        h = ((h ^ byte) * FNV_PRIME) & _MASK
    return h


def fnv1a_ints_fast(arr: np.ndarray, seed: int = FNV_OFFSET) -> int:
    """Same digest, one pass over the little-endian byte view.

    Identical by construction: ``int32.tobytes()`` on a little-endian build is
    the same byte sequence the JS loop emits by hand. Only used where speed
    matters; the checksum chain test runs both and compares.
    """
    return fnv1a_bytes(np.ascontiguousarray(arr, dtype="<i4").tobytes(), seed)


def fnv1a_string(text: str, seed: int = FNV_OFFSET) -> int:
    """FNV-1a over the UTF-16 code units of a string, byte-wise little-endian.

    JS iterates ``charCodeAt``, i.e. UTF-16 code units, so an astral character
    contributes two surrogates. Encoding to UTF-16-LE reproduces that exactly,
    where iterating Python code points would not.
    """
    h = seed & _MASK
    data = text.encode("utf-16-le")
    for i in range(0, len(data), 2):
        h = ((h ^ data[i]) * FNV_PRIME) & _MASK
        h = ((h ^ data[i + 1]) * FNV_PRIME) & _MASK
    return h


def to_hex32(h: int) -> str:
    """Unsigned 32-bit int -> lowercase 8-digit hex."""
    return f"{h & _MASK:08x}"


def checksum_ints(ints) -> str:
    return to_hex32(fnv1a_ints(ints))
